12-offer-encoding.md

```diff
@@ -0,0 +1,1056 @@
+
+Each form is signed using one or more TLV signature elements; TLV
+types 240 through 1000 are considered signature elements.  For these
+the tag is "lightning" || `messagename` || `fieldname`, and `msg` is the
+Merkle-root; "lightning" is the literal 9-byte ASCII string,
+`messagename` is the name of the TLV stream being signed (i.e. "offer", "invoice_request" or "invoice") and the `fieldname` is the TLV field containing the
+signature (e.g. "signature" or "payer_signature").
+
+The formulation of the Merkle tree is similar to that proposed in
+[BIP-taproot], with each TLV leaf paired with a nonce leaf to avoid
+revealing adjacent nodes in proofs (assuming there is a non-revealed TLV
+which has enough entropy).
+
+The Merkle tree's leaves are, in TLV-ascending order for each tlv:
+1. The H(`LnLeaf`,tlv).
+2. The H(`LnAll`||all-tlvs,tlv) where "all-tlvs" consists of all non-signature TLV entries appended in ascending order.
+
+The Merkle tree inner nodes are H(`LnBranch`, lesser-SHA256||greater-SHA256);
+this ordering means that proofs are more compact since left/right is
+inherently determined.
+
+If there are not exactly a power of 2 leaves, then the tree depth will
+be uneven, with the deepest tree on the lowest-order leaves.
+
+e.g. consider the encoding of an `offer` `signature` with TLVs TLV1, TLV2 and TLV3:
+
+```
+L1=H("LnLeaf",TLV1)
+L1nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV1) 
+L2=H("LnLeaf",TLV2)
+L2nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV2) 
+L3=H("LnLeaf",TLV3)
+L3nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV3) 
+
+Assume L1 < L1nonce, and L2 > L2nonce.
+
+   L1    L1nonce                      L2   L2nonce                L3   L3nonce
+     \   /                             \   /                       \   /
+      v v                               v v                         v v
+L1A=H("LnBranch",L1||L1nonce) L2A=H("LnBranch",L2nonce||L2)  L3A=H("LnBranch",L3nonce||L3)
+                 
+Assume L1A < L2A:
+
+       L1A   L2A                                 L3A=H("LnBranch",L3nonce||L3)
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightningoffersignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice: readers will either request an invoice
+(or multiple) or send an invoice based on the offer.  An offer can be much longer-lived than a
+particular invoice, so has some different characteristics; in particular it
+can be recurring, and the amount can be in a non-lightning currency.  It's
+also designed for compactness, to easily fit inside a QR code.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
```

Comment:
Would it be useful to add a field for a (list of) network addresses as a last resort to contact the node if the node is private and all route hints failed?

Reply:
I don't think it would be. Even in the case where the merchant doesn't care about his privacy it only allows us to contact the merchant but not to pay him. Because offers and onion messages are only a communication layer to get an invoice but in the end to send an actual payment we need to find a path in the channels graph and a ip address won't help us.

Reply:
The idea is that by contacting the merchant directly on an ip address and asking for an invoice, that invoice will contain fresh route hints that allows us to find that path.